File: src/main/java/mnm/mods/tabbychat/extra/filters/GuiFilterEditor.java
package mnm.mods.tabbychat.extra.filters;

import static mnm.mods.tabbychat.util.Translation.*;

import com.google.common.base.Joiner;
import com.google.common.base.Splitter;
import mnm.mods.tabbychat.api.filters.FilterSettings;
import mnm.mods.util.Color;
import mnm.mods.util.gui.GuiButton;
import mnm.mods.util.gui.GuiCheckbox;
import mnm.mods.util.gui.GuiComponent;
import mnm.mods.util.gui.GuiGridLayout;
import mnm.mods.util.gui.GuiLabel;
import mnm.mods.util.gui.GuiPanel;
import mnm.mods.util.gui.GuiText;
import net.minecraft.client.resources.I18n;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.SoundEvent;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.util.text.TextComponentTranslation;
import net.minecraft.util.text.TextFormatting;
import net.minecraftforge.fml.common.registry.GameRegistry;
import org.lwjgl.glfw.GLFW;

import javax.annotation.Nullable;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class GuiFilterEditor extends GuiPanel {

    private class ToggleButton extends GuiButton {
        private boolean active;

        private ToggleButton(String text) {
            super(text);
        }

        @Override
        public String getText() {
            String text = super.getText();
            TextFormatting color = active ? TextFormatting.GREEN : TextFormatting.RED;
            return color + text;
        }

        @Override
        public void onClick(double mouseX, double mouseY) {
            active ^= true;
        }
    }

    private UserFilter filter;
    private Consumer<UserFilter> consumer;

    private GuiText txtName;
    private GuiCheckbox chkRemove;
    private GuiText txtDestinations;
    private GuiCheckbox chkSound;
    private GuiText txtSound;
    private GuiText txtPattern;
    private GuiLabel lblError;

    private ToggleButton btnRegexp;
    private ToggleButton btnIgnoreCase;
    private ToggleButton btnRaw;

    public GuiFilterEditor(UserFilter filter, Consumer<UserFilter> consumer) {
        this.setLayout(new GuiGridLayout(20, 15));
        this.filter = filter;
        this.consumer = consumer;

        String pattern = filter.getRawPattern();
        FilterSettings settings = filter.getSettings();

        int pos = 0;

        this.addComponent(new GuiLabel(new TextComponentTranslation(FILTER_TITLE)), new int[]{8, pos, 1, 2});

        pos += 2;
        this.addComponent(new GuiLabel(new TextComponentTranslation(FILTER_NAME)), new int[]{1, pos});
        this.addComponent(txtName = new GuiText(), new int[]{5, pos, 10, 1});
        txtName.setValue(filter.getName());

        pos += 2;
        this.addComponent(new GuiLabel(new TextComponentTranslation(FILTER_DESTINATIONS)), new int[]{1, pos});
        this.addComponent(txtDestinations = new GuiText(), new int[]{8, pos, 10, 1});
        txtDestinations.setValue(Joiner.on(", ").join(settings.getChannels()));
        txtDestinations.setCaption(new TextComponentTranslation(FILTER_DESTIONATIONS_DESC));

        pos += 1;
        this.addComponent(btnRegexp = new ToggleButton(".*"), new int[]{1, pos, 2, 1});
        btnRegexp.active = filter.getSettings().isRegex();
        btnRegexp.setCaption(new TextComponentTranslation(FILTER_REGEX));
        this.addComponent(btnIgnoreCase = new ToggleButton("Aa"), new int[]{3, pos, 2, 1});
        btnIgnoreCase.active = settings.isCaseInsensitive();
        btnIgnoreCase.setCaption(new TextComponentTranslation(FILTER_IGNORE_CASE));
        this.addComponent(btnRaw = new ToggleButton("&0"), new int[]{5, pos, 2, 1});
        btnRaw.active = settings.isRaw();
        btnRaw.setCaption(new TextComponentTranslation(FILTER_RAW_INPUT));

        pos += 2;
        this.addComponent(new GuiLabel(new TextComponentTranslation(FILTER_HIDE)), new int[]{2, pos});
        this.addComponent(chkRemove = new GuiCheckbox(), new int[]{1, pos});
        chkRemove.setValue(settings.isRemove());

        pos += 1;
        this.addComponent(new GuiLabel(new TextComponentTranslation(FILTER_AUDIO_NOTIFY)), new int[]{2, pos});
        this.addComponent(chkSound = new GuiCheckbox(), new int[]{1, pos});
        chkSound.setValue(settings.isSoundNotification());

        pos += 1;
        this.addComponent(txtSound = new GuiText() {
            private int pos;

            @Override
            public boolean charTyped(char c, int key) {
                final int max = 10;
                if (key == GLFW.GLFW_KEY_DOWN) {
                    pos++;
                } else if (key == GLFW.GLFW_KEY_UP) {
                    pos--;
                }
                // suggest sounds
                final String val = getValue().toLowerCase()
                        .substring(0, getTextField().getCursorPosition());
                List<String> list = GameRegistry.findRegistry(SoundEvent.class).getKeys().stream()
                        .map(Object::toString)
                        .filter(s -> s.contains(val))
                        .collect(Collectors.toList());

                pos = Math.min(pos, list.size() - max);
                pos = Math.max(pos, 0);
                if (list.size() > max) {
                    list = list.subList(pos, pos + max);
                }
                setHint(Joiner.on('\n').join(list));
                if ((key == GLFW.GLFW_KEY_ENTER || key == GLFW.GLFW_KEY_KP_ENTER) && !list.isEmpty()) {
                    setValue(list.get(0));
                    GuiFilterEditor.this.setFocused(null);
                }
                return super.charTyped(c, key);
            }
        }, new int[]{3, pos, 14, 1});
        txtSound.setValue(settings.getSoundName().orElse(""));

        GuiButton play = new GuiButton("\u25b6") {
            @Override
            public SoundEvent getSound() {
                return GameRegistry.findRegistry(SoundEvent.class).getValue(new ResourceLocation(txtSound.getValue()));
            }
        };
        this.addComponent(play, new int[]{18, pos, 2, 1});

        pos += 2;
        this.addComponent(new GuiLabel(new TextComponentTranslation(FILTER_EXPRESSION)), new int[]{1, pos});
        this.addComponent(txtPattern = new GuiText() {
            @Override
            public boolean charTyped(char c, int key) {
                boolean r = super.charTyped(c, key);
                setPrimaryColor(Color.WHITE);
                lblError.setText(null);
                if (btnRegexp.active) {
                    // check valid regex
                    try {
                        filter.testPattern(getValue());
                    } catch (UserFilter.UserPatternException e) {
                        setPrimaryColor(Color.RED);
                        String string = e.getCause().getLocalizedMessage();
                        lblError.setText(new TextComponentString(string));
                    }
                }
                return r;
            }
        }, new int[]{8, pos, 12, 1});

        txtPattern.setValue(pattern == null ? "" : pattern);

        pos++;
        this.addComponent(lblError = new GuiLabel(), new int[]{4, pos});

        GuiButton accept = new GuiButton(I18n.format("gui.done")) {
            @Override
            public void onClick(double mouseX, double mouseY) {
                accept();
            }
        };
        this.addComponent(accept, new int[]{5, 14, 4, 1});

        GuiButton cancel = new GuiButton(I18n.format("gui.cancel")) {
            @Override
            public void onClick(double mouseX, double mouseY) {
                cancel();
            }
        };
        this.addComponent(cancel, new int[]{1, 14, 4, 1});
    }

    private void accept() {
        filter.setName(txtName.getValue());
        filter.setPattern(txtPattern.getValue());
        FilterSettings sett = filter.getSettings();
        sett.getChannels().clear();
        sett.getChannels().addAll(Splitter.on(",")
                .omitEmptyStrings()
                .trimResults()
                .splitToList(txtDestinations.getValue()));
        sett.setRemove(chkRemove.getValue());
        sett.setCaseInsensitive(btnIgnoreCase.active);
        sett.setRegex(btnRegexp.active);
        sett.setRaw(btnRaw.active);

        sett.setSoundNotification(chkSound.getValue());
        sett.setSoundName(txtSound.getValue());

        consumer.accept(filter);
        cancel();
    }

    private void cancel() {
        getParent().ifPresent(p -> p.setOverlay(null));
    }

    @Nullable
    @Override
    public GuiComponent getFocused() {
        return txtPattern;
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/extra/filters/GuiFilterEditor.kt
package mnm.mods.tabbychat.extra.filters
//
//import mnm.mods.tabbychat.client.gui.Label
//import mnm.mods.tabbychat.client.gui.widget.FieldBackedCheckboxButton
//import mnm.mods.tabbychat.util.Color
//import mnm.mods.tabbychat.util.Translation
//import mnm.mods.tabbychat.util.toComponent
//import net.minecraft.client.gui.FocusableGui
//import net.minecraft.client.gui.IGuiEventListener
//import net.minecraft.client.gui.IRenderable
//import net.minecraft.client.gui.widget.TextFieldWidget
//import net.minecraft.client.gui.widget.button.AbstractButton
//import net.minecraft.client.gui.widget.button.Button
//import net.minecraft.client.gui.widget.button.CheckboxButton
//import net.minecraft.client.resources.I18n
//import net.minecraft.util.ResourceLocation
//import net.minecraft.util.SoundEvent
//import net.minecraft.util.text.TextFormatting
//import net.minecraftforge.fml.common.registry.GameRegistry
//import net.minecraftforge.registries.ForgeRegistries
//import org.lwjgl.glfw.GLFW
//
//class GuiFilterEditor(private val filter: UserFilter) : FocusableGui(), IRenderable {
//
//    private val txtName: TextFieldWidget
//    private val chkRemove: CheckboxButton
//    private val txtDestinations: TextFieldWidget
//    private val chkSound: CheckboxButton
//    private val txtSound: TextFieldWidget
//    private val txtPattern: TextFieldWidget
//    private val lblError: Label
//
//    private val btnRegexp: ToggleButton
//    private val btnIgnoreCase: ToggleButton
//    private val btnRaw: ToggleButton
//
//    private val renderer = ArrayList<IRenderable>()
//    private val children = ArrayList<IGuiEventListener>()
//
//    override fun children() = children
//
//    internal inner class ToggleButton(
//            x: Int, y: Int,
//            w: Int, h: Int,
//            private val text: String
//    ) : AbstractButton(x, y, w, h, text) {
//
//        var toggle = false
//
//        init {
//            message = "${TextFormatting.RED}${text}"
//        }
//
//        private fun updateMessage() {
//            val color = if (toggle) TextFormatting.GREEN else TextFormatting.RED
//            message = color.toString() + text;
//        }
//
//        override fun onPress() {
//            toggle = !toggle
//            updateMessage()
//        }
//    }
//
//    init {
//        this.layout = GuiGridLayout(20, 15)
//
//        val pattern = filter.rawPattern
//        val settings = filter.settings
//
//        var pos = 0
//
//        this.add(Label(Translation.FILTER_NAME.toComponent()), intArrayOf(1, pos))
//        txtName = this.add(TextFieldWidget(), intArrayOf(5, pos, 10, 1)) {
//            value = filter.name
//        }
//
//        pos += 2
//        this.add(Label(Translation.FILTER_DESTINATIONS.toComponent()), intArrayOf(1, pos))
//        txtDestinations = this.add(TextFieldWidget(), intArrayOf(8, pos, 10, 1)) {
//            value = settings.channels.joinToString(", ")
////            caption = Translation.FILTER_DESTIONATIONS_DESC.toComponent()
//        }
//
//        pos += 1
//        btnRegexp = this.add(ToggleButton(".*"), intArrayOf(1, pos, 2, 1)) {
//            toggle = filter.settings.isRegex
////            caption = Translation.FILTER_REGEX.toComponent()
//        }
//        btnIgnoreCase = this.add(ToggleButton("Aa"), intArrayOf(3, pos, 2, 1)) {
//            toggle = settings.isCaseInsensitive
////            caption = Translation.FILTER_IGNORE_CASE.toComponent()
//        }
//        btnRaw = this.add(ToggleButton("&0"), intArrayOf(5, pos, 2, 1)) {
//            toggle = settings.isRaw
////            caption = Translation.FILTER_RAW_INPUT.toComponent()
//        }
//
//        pos += 2
//        this.add(Label(Translation.FILTER_HIDE.toComponent()), intArrayOf(2, pos))
//        chkRemove = this.add(FieldBackedCheckboxButton(), intArrayOf(1, pos)) {
//            value = settings.isRemove
//        }
//
//        pos += 1
//        this.add(Label(Translation.FILTER_AUDIO_NOTIFY.toComponent()), intArrayOf(2, pos))
//        chkSound = this.add(FieldBackedCheckboxButton(), intArrayOf(1, pos)) {
//            value = settings.isSoundNotification
//        }
//
//        pos += 1
//        val play = this.add(Button("\u25b6"), intArrayOf(18, pos, 2, 1))
//        txtSound = this.add<TextFieldWidget>(object : TextFieldWidget() {
//            private var index: Int = 0
//
//            override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
//                val max = 10
//                if (keyCode == GLFW.GLFW_KEY_DOWN) {
//                    index++
//                } else if (keyCode == GLFW.GLFW_KEY_UP) {
//                    index--
//                }
//                // suggest sounds
//                val value = value.toLowerCase().substring(0, delegate.cursorPosition)
//                var list = GameRegistry.findRegistry(SoundEvent::class.java).keys.asSequence()
//                        .map { it.toString() }
//                        .filter { s -> s.contains(value) }
//                        .toList()
//
//                index = index.coerceAtMost(list.size - max)
//                index = index.coerceAtLeast(0)
//                if (list.size > max) {
//                    list = list.subList(index, index + max)
//                }
//                hint = list.joinToString("\n")
//
//                if ((keyCode == GLFW.GLFW_KEY_ENTER || keyCode == GLFW.GLFW_KEY_KP_ENTER) && list.isNotEmpty()) {
//                    this.value = list[0]
//                    setFocused(null)
//                }
//                return super.keyPressed(keyCode, scanCode, modifiers)
//            }
//        }, intArrayOf(3, pos, 14, 1)) {
//            value = settings.soundName ?: ""
//            delegate.maxStringLength = 100
//            delegate.setValidator { txt -> ResourceLocation.tryCreate(txt) != null }
//            delegate.setResponder { s ->
//                val res = ResourceLocation.tryCreate(s)
//                play.sound = ForgeRegistries.SOUND_EVENTS.getValue(res)
//            }
//        }
//        lblError = Label()
//
//        pos += 2
//        this.add(Label(Translation.FILTER_EXPRESSION.toComponent()), intArrayOf(1, pos))
//        txtPattern = this.add(TextFieldWidget(), intArrayOf(8, pos, 12, 1)) {
//            value = pattern
//            delegate.setResponder {
//                primaryColor = Color.WHITE
//                lblError.text = null
//                if (btnRegexp.active) {
//                    // check valid regex
//                    try {
//                        filter.testPattern(it)
//                    } catch (e: UserFilter.UserPatternException) {
//                        primaryColor = Color.RED
//                        lblError.text = e.cause?.localizedMessage?.toComponent()
//                    }
//
//                }
//            }
//        }
//
//        pos++
//        this.add(lblError, intArrayOf(4, pos))
//
//        this.add(Button(I18n.format("gui.done")) {
//            accept()
//        }, intArrayOf(5, 14, 4, 1))
//    }
//
//    override fun render(p0: Int, p1: Int, p2: Float) {
//        TODO("Not yet implemented")
//    }
//
//    private fun accept() {
//        filter.apply {
//            name = txtName.value
//            setPattern(txtPattern.value)
//            settings.apply {
//                channels = txtDestinations.value.split(",").filter { it.isNotBlank() }.toMutableList()
//                isRemove = chkRemove.value
//                isCaseInsensitive = btnIgnoreCase.toggle
//                isRegex = btnRegexp.toggle
//                isRaw = btnRaw.toggle
//
//                isSoundNotification = chkSound.value
//                soundName = txtSound.value
//            }
//        }
//    }
//}
